Integrate CLI with DPPM proper; add user/group modification commands

dscottboggs opened the pull request.

dscottboggs committed:
added user modification CLI

j8r commented on src/cli.cr:
```diff
@@ -40,6 +167,179 @@ CLI.run(
+
+            Ideally, you won't have multiple users with the same name, or you
+            can use the web interface to edit these values using API keys for
+            exact identification.
```
Absolutely too long, users won't want to have their terminals taken with textual information.
This documentation has its place in a `man` or a proper documentation place (web, markdown...)

dscottboggs replied:
Updated as requested

j8r commented on src/cli.cr:
```diff
@@ -40,6 +167,179 @@ CLI.run(
+                commas, and multiple group selectors by separating
+                them with a colon, like `match-groups=1,2,3:2,4', which
+                would select a user who's a member of groups 1, 2, and
+                3, or a user who's a member of groups 2 and 4.
```
Same here, way too long

j8r commented on src/cli.cr:
```diff
@@ -40,6 +167,179 @@ CLI.run(
+
+            Ideally, you won't have multiple users with the same name, or you
+            can use the web interface to edit these values using API keys for
+            exact identification.
```
Way too long. Look at other clis in your system as examples.

j8r reviewed (changes requested):
The cli information should be a short description of the command, option or variable. Else, the terminal is filled with too much text.
If the user doesn't know what a command do, (s)he can refer to the docs where details are explained.

dscottboggs committed:
Added group editing abilities

j8r commented on src/cli.cr:
```diff
@@ -40,6 +231,235 @@ CLI.run(
+                       member of groups 2 and 4."
+              },
+              api_key: {
+                short: "key",
```
Why having this? `short` is only useful for options, usually a char.

dscottboggs committed:
simplify --help text; update readme

parent 79540ec8ee286aa4013cb6a56dd4c6a130b80df4
author D. Scott Boggs <scott@tams.tech> 1557696343 -0400
committer D. Scott Boggs <scott@tams.tech> 1557859035 -0400

Revise CLI variable structure

Syncing work from laptop

revised -- compiles with requested changes

move clicr macro into the DppmRestApi::CLI namespace

Minor revisions

WIP: specs for CLI methods

syncing work from laptop

Moar specs

Not sure why this spec fails; remove pp!s

Sketch of reworking Config API

Update to reflect changes in dependency

j8r commented on src/cli.cr:
```diff
@@ -40,6 +256,180 @@ CLI.run(
+              },
+              key:      {info: "the query key to match for"},
+              add_glob: {
+                short: "add",
```
Useless

j8r commented on src/cli.cr:
```diff
@@ -40,6 +256,180 @@ CLI.run(
+                info: "The groups that selected users must be a member of.",
+              },
+              api_key: {
+                short: "key",
```
Useless

j8r commented on src/cli.cr:
```diff
@@ -40,6 +256,180 @@ CLI.run(
+                info:  "the api key to select a user by.",
+              },
+              output_file: {
+                short: "file",
```
Ditto

j8r commented on src/cli.cr:
```diff
@@ -40,6 +256,180 @@ CLI.run(
+                info: "The groups that selected users must be a member of.",
+              },
+              api_key: {
+                short: "key",
```
Ditto

j8r commented on src/cli.cr:
```diff
@@ -40,6 +256,180 @@ CLI.run(
+                info: "The groups that selected users must be a member of.",
+              },
+              api_key: {
+                short: "key",
```
Ditto

j8r reviewed (dismissed):
The variables are applied recursively, you don't need to copy yourself.

j8r commented:
Trailing dots can also be removed, they aren't commonly present for short descriptions.

j8r commented:
`Clicr` updated to `0.6.0`, you can now use `description` with heredocs for long descriptions.
As you wish if you prefer to keep this information in the README of to put them in the cli.

j8r commented:
Resolves https://github.com/DFabric/dppm-rest-api/issues/4

j8r commented on src/cli.cr:
```diff
@@ -40,6 +270,149 @@ CLI.run(
+              path: {
+                info: "the path on which to update allowed query parameters.",
+              },
+              key:      {info: "the query key to match for"},
```
put it in a multi line

j8r commented on src/cli.cr:
```diff
@@ -40,6 +270,149 @@ CLI.run(
+                  "/{app,pkg,src,service}/**" => DppmRestApi::Config::Route.new(
+                    DppmRestApi::Access::All,
+                    {"namespace" => ["default-namespace"]}),
+                }.to_json,
```
That's weird. Can't it be a constant somewhere, I suppose it can be reusable it other places?

dscottboggs replied:
Fair point

j8r commented on src/cli.cr:
```diff
@@ -40,6 +270,149 @@ CLI.run(
       },
+      add_user: {
+        info:      "add a user",
+        action:    "DppmRestApi::CLI.add_user",
```
`DppmRestApi::CLI.` can be dropped (see also all other places below)

dscottboggs replied:
It can just be `action: :add_user`?

dscottboggs replied:
I tried setting the values like this and it said `undefined method 'run_server' for CLI:Module`

j8r replied:
That depends where the macro is expanded.
I think it can be inside the `CLI` module, inside `def run`, then call `CLI.run`. I've pushed a commit to change `CLI` to `DPPM::CLI`.

dscottboggs replied:
I see. Ok.

dscottboggs replied:
This breaks things. See the difference between [8aac32c2 and c1c4e02](https://github.com/DFabric/dppm-rest-api/pull/8/commits/8aac32c2718b9ae0eb8c3f508386288774de3abd). I think that's an issue with DPPM's not being namespaced?

dscottboggs replied:
nevermind it works after running shards update

dscottboggs replied:
Actually it doesn't.

`action: :run_server` gives `undefined method 'run_server' for DPPM::CLI:Module`

`action: run_server` or `action: DppmRestApi::CLI.run_server` gives `undefined macro method 'Call#split'`

j8r reviewed (commented):
What is the `mn_`? Can't it be just `data_dir` and `groups`?
Or, you can have `raw_` appended to the variable to underline the fact it's a raw string.

j8r commented on src/cli.cr:
```diff
@@ -2,7 +2,80 @@ require "dppm/cli"
+  private def select_users(match_name, match_groups, api_key, from selected_users)
+    # Filter by name
+    if name = match_name
+      if name.starts_with?('/') && name.ends_with?('/')
```
```suggestion
      if stripped_name = name.lchop?('/').try &.rchop?('/')
```

dscottboggs replied:
*Very* nice. I had a feeling I didn't have the best solution.

j8r commented on src/cli.cr:
```diff
@@ -2,7 +2,80 @@ require "dppm/cli"
+    if name = match_name
+      if name.starts_with?('/') && name.ends_with?('/')
+        # match by regex
+        name_regex = Regex.new name.lchop('/').rchop('/')
```
```suggestion
        name_regex = Regex.new stripped_name
```

j8r commented on src/cli.cr:
```diff
@@ -2,7 +2,80 @@ require "dppm/cli"
+      selected_users = selected_users.select { |user| user.api_key_hash == key }
+      # freak out if multiple users were found.
+      if selected_users.size > 1
+        abort String.build { |msg|
```
Use `do ... end`

dscottboggs replied:
I used curly-braces in this case because it has different precedence than `do..end` I will save as an intermediate variable to avoid confusion

j8r commented on src/cli.cr:
```diff
@@ -13,6 +86,163 @@ module DppmRestApi::CLI
+    puts
+  end
+
+  def add_group(id mn_id, name mn_name, permissions mn_permissions, data_dir mn_data_dir, **args)
```
It would be simpler to have `add_group(id, name, permissions, data_dir, **args)`, the `String | Nil` -> `String` cast doesn't need an intermediate variable.

dscottboggs replied:
I don't get what you mean? Of course a cast from nil requires an intermediate variable?

j8r replied:
No:
```cr
name = if true
  ""
else
  nil
end
puts typeof(name)
name || raise "that's nil!"
puts typeof(name) #=> String
```

j8r replied:
No:
```cr
name = if true
  ""
else
  nil
end
puts typeof(name) #=> (String | Nil)
name || raise "that's nil!"
puts typeof(name) #=> String
```

dscottboggs replied:
woaah. I did not know that.

dscottboggs replied:
woaah. I did not know that.

> Please `raise` instead of `abort`, the exception is then catched with a nice colored `ERR! this is my error` logging.

Ah. :+1:

j8r commented:
Please `raise` instead of `abort`, the exception is then catched with a nice colored `ERR! this is my error` logging.

j8r commented on src/cli.cr:
```diff
@@ -2,7 +2,80 @@ require "dppm/cli"
+
+  # A helper method for the select_users method -- splits a comma-separated
+  # list of numbers in a string into an `Array` of `Int32` values.
+  private def split_groups_arg(argument text : String) : Array(Int32)
```
Why having space separated `argument text`? That's confusing. `numbers_list` is better.

dscottboggs replied:
agreed for sure

j8r commented on src/cli.cr:
```diff
@@ -2,7 +2,80 @@ require "dppm/cli"
+  # A helper method for the select_users method -- splits a comma-separated
+  # list of numbers in a string into an `Array` of `Int32` values.
+  private def split_groups_arg(argument text : String) : Array(Int32)
+    text.split(',').map do |g|
```
That's a detail, this is creating an intermediate array.
Using the block variant of `split`, then concatenating with `numbers << g_int` will be more efficient.

dscottboggs replied:
Thanks!

j8r commented on src/cli.cr:
```diff
@@ -2,7 +2,80 @@ require "dppm/cli"
+  # A helper method for the select_users method -- splits a comma-separated
+  # list of numbers in a string into an `Array` of `Int32` values.
+  private def split_groups_arg(argument text : String) : Array(Int32)
+    text.split(',').map do |g|
```
what is `g`? `raw_group_id` is a better name.

j8r commented on src/config.cr:
```diff
@@ -32,6 +32,25 @@ struct DppmRestApi::Config
+    # to prevent data loss in case of loss of power during the file write.
+    File.rename tmp_file.path, path.to_s
+  ensure
+    if tf = tmp_file
```
Not needed, this type of conditions are mainly needed when we deal with method calls, not variables

dscottboggs replied:
I'm sorry, which isn't needed? `if tf = tmp_file`?

j8r replied:
The assignement isn't needed, sorry. Just `if tmp_file`.

dscottboggs replied:
See, I thought that was the case, but someone in the chat said that there were circumstances where tmp_file could be non-nil in the conditional but then be nil when it gets to the call site, so you should just always assign to an intermediate variable. I just took their word for it -- no harm in doing it anyway, and I don't like `not_nil!` everywhere either.

Not that I'm saying I should keep it like this, just explaining why I did that -- it's just a habit I've developed. Now that you've brought it to my attention, though, I think `tmp_file.try &.delete` is better overall.

j8r commented on src/cli.cr:
```diff
@@ -2,7 +2,247 @@ require "dppm/cli"
+              action:    "DppmRestApi::CLI.edit_users",
+              alias:     "e",
+              variables: {
+                new_name:   {info: "The name to change the selected user's name to"},
```
Can be split to a multiline, like others

j8r commented on src/cli.cr:
```diff
@@ -2,7 +2,247 @@ require "dppm/cli"
+                },
+                remove_glob: {
+                  short: "rm",
+                  info:  "\
```
Why the backslash?

j8r commented on src/cli.cr:
```diff
@@ -2,7 +2,247 @@ require "dppm/cli"
+              alias:     "rm",
+              action:    "DppmRestApi::CLI.delete_group",
+              variables: {
+                group_id: {info: "The ID of the group to remove"},
```
Ditto, multiline

j8r commented on src/cli.cr:
```diff
@@ -2,7 +2,247 @@ require "dppm/cli"
+    # Filter by group membership
+    if groups = match_groups
+      group_matches = Set(Set(Int32)).new
+      groups.split(':').each do |part|
```
You can use the block variant directly

j8r reviewed (commented):
Looks good overall, missing a few tests

dscottboggs commented:
> missing a few tests

Good point!

j8r commented on src/config.cr:
```diff
@@ -1,20 +1,21 @@
   property users : Array(User)
 
-  def user(*, named username : String) : User?
+  def user_named(username : String) : User?
```
```suggestion
  def user_from_name(username : String) : User?
```

j8r commented on src/config.cr:
```diff
@@ -1,20 +1,21 @@
   end
 
-  def user(*, authenticated_with api_key : String)
+  def user_authenticated_with(api_key : String)
```
```suggestion
  def authenticated_user_from_api_key(api_key : String)
```

j8r replied:
A bit long, if you have a better idea be free.

j8r commented on src/config.cr:
```diff
@@ -1,20 +1,21 @@
   end
 
-  def group(*, named group_name : String) : Group?
+  def group_named(group_name : String) : Group?
```
```suggestion
  def group_from_name(group_name : String) : Group?
```

j8r commented on test.cr:
```diff
@@ -0,0 +1,8 @@
+       end
+puts typeof(name) # => (String | Nil)
+name || abort "that's nil!"
+puts typeof(name) # => String
```
?

dscottboggs replied:
It was never here *waves hand*
I had copy-pasted that snippet into a file to run it, and it must've accidentally gotten committed when I was syncing my work from one computer to another

j8r replied:
I do this sometimes too

j8r commented on spec/permissions.json:
```diff
@@ -1,66 +1,80 @@
       "permissions": {
         "/**": {
-          "permissions": "All"
+          "permissions": 15
```
What are all this `15`? It was nice before.

dscottboggs replied:
It must be the default to serialize them as the numeric value, I can look into changing that.

j8r commented on src/config.cr:
```diff
@@ -1,20 +1,21 @@
   end
 
-  def user(*, authenticated_with api_key : String)
+  def user_authenticated_by(api_key : String)
```
If we have a stack trace, with `user_authenticated_by` won't give us any clue what is it.
```suggestion
  def authenticated_user_from_api_key(api_key : String)
```
A bit long, but is it really a problem? Not sure we will use it that often.

dscottboggs replied:
It's not that it was a problem I just had done that before you suggested your version. I'll change it now so I don't forget.

dscottboggs replied:
Actually, the *only* place where it even *would* be used... it [isn't](https://github.com/DFabric/dppm-rest-api/blob/b8b514f40038009c177c5dab8279bce21f41a339/src/middlewares/auth_middleware.cr#L34). I think we should just remove these methods.

dscottboggs replied:
So I realize it's not working right now, but I wanted to get the idea fleshed out and see what you think. I think my Kemal Auth library can actually be reworked to be simpler, making use of modules as interfaces rather than needing a generic to be defined. I think I'm going to look into that before going on.

j8r commented on shard.yml:
```diff
@@ -12,7 +12,8 @@ dependencies:
     branch: master
   scrypt:
-    github: ysbaddaden/scrypt-crystal
+    github: dscottboggs/scrypt-crystal
```
Have you done a PR for this?

dscottboggs replied:
https://github.com/ysbaddaden/scrypt-crystal/pull/6

j8r commented on spec/actions/app_spec.cr:
```diff
@@ -28,37 +28,37 @@ module DppmRestApi::Actions::App
     describe (fmt_route "/:app_name/service/boot") do
       pending "responds with 401 Forbidden" do
-        put (pp! fmt_route "/some_app/service/boot")
+        put (fmt_route "/some_app/service/boot")
```
Please, focus on the PR title to keep changes as minimal as possible - integrate the CLI.

dscottboggs committed:
Spec all the CLI methods

Change Access serialization from numeric to string

dscottboggs committed:
Add test API keys to the repo

dscottboggs committed:
Add add_route CLI endpoint

Change access serialization from string to list

j8r commented on run-tests.sh:
```diff
@@ -0,0 +1,22 @@
+  build=yes
+fi
+
+set -e
```
Duplicated

j8r commented on run-tests.sh:
```diff
@@ -0,0 +1,22 @@
+#!/bin/sh
+
+set -e
```
I also often add `-u` (`set -eu`) to ensure all variables are defined

dscottboggs replied:
I meant to have `set -e` in one place, and `set -u` in another, to allow `$1` to be empty. Thanks for catching this

j8r replied:
A trick is to use `${1-}` to bypass this.

dscottboggs replied:
Cool

j8r commented on run-tests.sh:
```diff
@@ -0,0 +1,22 @@
+set -e
+
+build=no
+if [ "$1" = 'and' ] && [ "$2" = 'build' ]; then
```
What is `and`?

dscottboggs replied:
`$  ./run-tests and build`

Just being clever. I'll cut it down.

j8r replied:
That can be an option, `--build`.
I guess this script will be used in the CI, too?

dscottboggs replied:
It doesn't need to be, it's just there for my own convenience.

j8r commented on spec/permissions.json:
```diff
@@ -1,66 +1,80 @@
         "/**": {
-          "permissions": "All"
+          "permissions": "Create | Read | Update | Delete"
         }
```
Why this and not a list? This is a list of permissions.

dscottboggs replied:
The `Access` type can be serialized as any of

- its numeric value,
- the above single-string form, or
- an array of string values.

The reserialization can be any of the above. There's no particular reason I picked the above form, it's just the first form that accepts multiples when you're reading it back. If you prefer the array form it can be that.

j8r replied:
I think it's better be as close as standard json as possible, especially if  to be able to use external tools (`jq`...)

j8r replied:
I think it's better to be as close as standard json as possible, especially if  to be able to use external tools (`jq`...)

j8r replied:
I think it's better to be as close as standard json as possible, especially to be able to use external tools (`jq`...)

j8r commented on spec/normal_user.api_key:
```diff
@@ -0,0 +1 @@
+kP/U4BBQV0s2TcwN2MWwFrbwGDyMGkZEFt0era1yJc9eSuBhC+PiaDF1L+eylxsEoOzrnIsd4LgYOGQcIP6Y
```
Not sure it's needed to store this in a file.

dscottboggs replied:
Should it just be generated on the fly as a constant then?

j8r replied:
It will be better than reading to a file

dscottboggs committed:
Remove API key files (used for testing)

Test keys are now randomly generated for every spec run.

dscottboggs commented:
Can we merge this?

j8r commented:
The CI will tell, it is red.

dscottboggs commented:
Yes, I need to merge the changes to `Scrypt::Password` into this and the JWT auth shard, then it should be ok. Is doing a separate branch/PR for that, then rebasing the right thing to do?

j8r commented:
Using your fork was fine. It's temporary and not perfect, but at least everyone can compile the project.

dscottboggs commented:
Ugh if I did that I'd have to wait for that to be merged and it won't even fix every instance. I'll just fix it on this branch and in the future I will create smaller-scale and more atomic PRs

dscottboggs committed:
remove deprecated method from Scrypt::Password

dscottboggs commented:
Review is required for merging

j8r commented on run-tests.sh:
```diff
@@ -0,0 +1,21 @@
+ameba
+KEMAL_ENV=test crystal spec $@
+[ $build = yes ] && do_build
+echo All OK
```
Missing new line
